src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs

```diff
@@ -42,15 +42,79 @@ public IISExpressDeployer(DeploymentParameters deploymentParameters, ILoggerFact
                 // Start timer
                 StartTimer();
 
-                // For now we always auto-publish. Otherwise we'll have to write our own local web.config for the HttpPlatformHandler
-                DotnetPublish();
+                var targetFramework = DeploymentParameters.TargetFramework;
 
-                var contentRoot = DeploymentParameters.PublishedApplicationRootPath;
+                // IISIntegration uses this layout
+                var dllRoot = Path.Combine(DeploymentParameters.ApplicationPath, "bin", DeploymentParameters.RuntimeArchitecture.ToString(),
+                    DeploymentParameters.Configuration, targetFramework);
+
+                if (!Directory.Exists(dllRoot))
+                {
+                    // Most repos use this layout
+                    dllRoot = Path.Combine(DeploymentParameters.ApplicationPath, "bin", DeploymentParameters.Configuration, targetFramework);
+
+                    if (!Directory.Exists(dllRoot))
+                    {
+                        // The bits we need weren't pre-compiled, compile on publish
+                        DeploymentParameters.PublishApplicationBeforeDeployment = true;
+                    }
+                    else if (DeploymentParameters.RuntimeFlavor == RuntimeFlavor.Clr
+                        && DeploymentParameters.RuntimeArchitecture == RuntimeArchitecture.x86)
+                    {
+                        // x64 is the default. Publish to rebuild for the right bitness
+                        DeploymentParameters.PublishApplicationBeforeDeployment = true;
+                    }
+                }
+
+                if (DeploymentParameters.RuntimeFlavor == RuntimeFlavor.CoreClr
+                        && DeploymentParameters.ApplicationType == ApplicationType.Standalone)
+                {
+                    // Publish is always required to get the correct standalone files in the output directory
+                    DeploymentParameters.PublishApplicationBeforeDeployment = true;
+                }
+
+                var contentRoot = string.Empty;
+                if (DeploymentParameters.PublishApplicationBeforeDeployment)
+                {
+                    DotnetPublish();
```

Comment:
This will publish to a temp folder. Is that ever cachable between tests that aren't with a fixture?

Reply:
Not yet.